Wire up Data Service to Scheduling Manager using JSON-RPC

wjbuys opened the pull request:
This deploys the DataService as a lambda function that accepts JSON-RPC payloads, and adds a JSON-RPC client to the Scheduling Manager in order to call the DataService.

Right now, this wires up the DataService's Lambda function into the scheduler using Cloudformation's ImportValue function. While this is nice for development, we'll probably want to figure out a better long term way to do this.

Implements #227

#### Testing
So far, testing has been manual. After running `gradle deploy` to deploy the full stack, you have to manually create an Environment that targets a cluster that your account owns, and then wait for a minutely reconciliation run.

- [ ] TODO: Add end-to-end test that does this

#### Licensing

This contribution is under the terms of the Apache 2.0 License: Yes

Commit:
Add JSON-RPC-over-lambda client and server implementations

In order to use a single Lambda function for the DataService, we need
some form of RPC over the Lambda input/output streams. The jsonrpc4j
project supports this pretty well.

- `json-rpc-lambda-client` wraps a LambdaAsyncClient to call a lambda
  function with JSON-RPC formatted request and response payloads. It
  uses jsonrpc4j's Proxy support to dynamically generate a client for
  any given Java interface.
- `json-rpc-lambda-server` implements `RequestStreamHandler` and plumbs
  it into the Input-/OutputStream based JsonRpcServer.

This allows us to transparently call methods on a Java interface and
have those translated into calls into a lambda function.

This means that if it turns out JSON-RPC isn't a good fit for the
protocol, we can replace this transport layer, and our Java interfaces
can remain unchanged.

Since these are fairly thin wrappers, they're pretty hard to test. I
think it's okay to rely on DatService integration tests for now.

Commit:
Deploy the DataService as a lambda function, wire into scheduler

This deploys the DataService as a lambda function that accepts JSON-RPC
payloads, and adds a JSON-RPC client to the Scheduling Manager in order
to call the DataService.

Right now, this wires up the DataService's Lambda function into the
scheduler using Clouformation's ImportValue function. While this is nice
for development, we'll probably want to figure out a better long term
way to do this.

Implements #227

wjbuys commented:
I've been trying to set up the end-to-end test for this, and I'm running into issues re-running the tests. In particular, if we don't clear out the data in DDB, then we'll keep scheduling more and more tasks for each new test environment created. I think it might be generally useful to have an easy way to clear out all the state from the test stack, so I'm going to add a utility for that to our end-to-end-tests package.

narehayrapetyan commented:
Yes, I agree, the tests should clean up after themselves.

narehayrapetyan commented on json-rpc-lambda-client/build.gradle:
```diff
@@ -0,0 +1,18 @@
+    // we don't have to depend on a specific SDK version here:
+    compile "software.amazon.awssdk:lambda:+"
+
+    // For @Slf4j annotation
```
why only for slf4j? won't this comment become out of date when we use lombok for other stuff?

wjbuys replied:
Sure.

narehayrapetyan commented on json-rpc-lambda-client/src/main/java/com/amazonaws/blox/clients/JsonRpcLambdaClient.java:
```diff
@@ -0,0 +1,193 @@
+    this(defaultObjectMapper(), lambda, functionName);
+  }
+
+  private JsonRpcLambdaClient(ObjectMapper mapper, LambdaAsyncClient lambda, String functionName) {
```
what's the advantage of having two constructors? won't it be more concise to autowire the dependencies?

wjbuys replied:
This was a refactoring step from when the constructor didn't have an ObjectMapper parameter. Will remove the redundant constructor.

narehayrapetyan commented on json-rpc-lambda-client/src/main/java/com/amazonaws/blox/clients/JsonRpcLambdaClient.java:
```diff
@@ -0,0 +1,193 @@
+   * @param <T>
+   * @return
+   */
+  public static <T> T newProxy(
```
why not move this outside of this class so we don't have a mix of instantiated and static methods? it looks like some methods are calling the same methods on the client as well, it would help with readability imho to separate these out

narehayrapetyan replied:
looks like renaming the client to jsonRPCclient or similar would make it clearer

wjbuys replied:
I was trying to contain the this implementation detail to as small a surface as possible. After our discussion, I think splitting this off into a separate class, and making the JsonRpcLambdaClient package-local will achieve that more cleanly.

narehayrapetyan commented on json-rpc-lambda-client/src/main/java/com/amazonaws/blox/clients/JsonRpcLambdaClient.java:
```diff
@@ -0,0 +1,193 @@
+    this(defaultObjectMapper(), lambda, functionName);
+  }
+
+  private JsonRpcLambdaClient(ObjectMapper mapper, LambdaAsyncClient lambda, String functionName) {
```
why have 2 different constructors? and why not autowire?

narehayrapetyan commented on json-rpc-lambda-client/src/main/java/com/amazonaws/blox/clients/JsonRpcLambdaClient.java:
```diff
@@ -0,0 +1,193 @@
+      LambdaAsyncClient lambdaClient,
+      String functionName) {
+    IJsonRpcClient client = new JsonRpcLambdaClient(mapper, lambdaClient, functionName);
+    return ProxyUtil.createClientProxy(
```
from the documentation it looks like this can be autowired?

wjbuys replied:
As we discussed, I'm trying to localize all of the unpleasant glue for this in one place, so that we don't have multiple classes that refer to the JSON-RPC implementation details spread through our application. I'll add the explanatory javadoc for that.

narehayrapetyan commented on json-rpc-lambda-client/src/main/java/com/amazonaws/blox/clients/JsonRpcLambdaClient.java:
```diff
@@ -0,0 +1,193 @@
+  }
+
+  /** {@inheritDoc} */
+  @Override
```
what's this overriding? (and the rest) we're just implementing an interface no?

wjbuys replied:
As discussed, these are unfortunately required generic wrappers from the `IJsonRpcClient`interface for the corresponding methods that return `Object`. I'll replace the inheritdoc comment with an explanation.

narehayrapetyan reviewed (commented):
So are the scheduling functions using this already?

Commit:
Refactor JsonRpcLambdaClient to cleanly separate out glue code

This moves the implementation of the IJsonRpcClient class out of the
main JsonRpcLambdaClient, and adds JavaDoc to clarify what each `invoke`
overload of the client does.

Commit:
Remove unhelpful comment

wjbuys commented:
> So are the scheduling functions using this already?

Yes, [this](https://github.com/blox/blox/pull/281/files#diff-1d1f6dc21c6be2d5d91e3a504dbf93a0R79) is what wires it up.
